### PR TITLE
Added unit test for DeckPicker with No Permissions

### LINKED ISSUE #7849: Unit Test: Ensure that Deck Picker works without permissions
Also needs a follow-up issue to get a unit test in, or a drive to get the instrumented tests working on a clean/safe collection

_Originally posted by @david-allison-1 in https://github.com/ankidroid/Anki-Android/issues/7846#issuecomment-741636510_

### PR DESCRIPTION
**Pull request by shobhi1310**
## Purpose / Description
Automated testing such that developers can't accidentally make changes which cause AnkiDroid to crash if it doesn't have storage permission.

## Fixes
Fixes #7849 

## Approach
Create an instance of the app using shadow and remove file access permissions both read and write. Then create and instance of `deckPicker` class and run it, expected result should be that it again calls `deckPicker` class as it will ask user to grant permission.

## How Has This Been Tested?
It is a unit test.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by Arthur-Milchior** [comment]
The test fails on github. I assume you did run it on your computer. In this case, I'm sorry, there is some difference here, I don't know what

> com.ichi2.anki.DeckPickerTest > deckPickerNotCrashOnNoPermissionTest FAILED
    java.lang.NullPointerException
        at com.ichi2.anki.DeckPickerTest.lambda$deckPickerNotCrashOnNoPermissionTest$6(DeckPickerTest.java:269)
        at androidx.test.core.app.ActivityScenario.lambda$onActivity$2$ActivityScenario(ActivityScenario.java:660)
        at androidx.test.core.app.ActivityScenario$$Lambda$4.run(Unknown Source)
        at androidx.test.core.app.ActivityScenario.onActivity(ActivityScenario.java:670)
        at com.ichi2.anki.DeckPickerTest.deckPickerNotCrashOnNoPermissionTest(DeckPickerTest.java:266)

**Comment by shobhi1310** [comment]
@david-allison-1 

I am not able to deny the permissions to read and write storage in the testing part. As by default also it is saying that it does not have permissions still, `getCol().getDecks().count()` is printing 1. Not like the actual scenario in app.

**Comment by david-allison-1** [comment]
> I am not able to deny the permissions to read and write storage in the testing part. As by default also it is saying that it does not have permissions still, `getCol().getDecks().count()` is printing 1. Not like the actual scenario in app.

You should be able to use code such as this to allow or deny permissions:

https://github.com/ankidroid/Anki-Android/blob/bd639ad715821df2c54f962c983ba4a45b39eefa/AnkiDroid/src/test/java/com/ichi2/anki/InitialActivityTest.java#L75-L79

> `colIsOpen()` is also showing true though the permissions are denied. Why is it so?

Unsure - you'll need to go in with the debugger to verify. If you need me to deep-dive, let me know

**Comment by shobhi1310** [comment]
@david-allison-1
Could you say what might have gone wrong in the approach as one of the test has failed.

File: AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java [hunk 1]
```diff
@@ -246,6 +261,25 @@ public void databaseLockedNoPermissionIntegrationTest() {
         }
     }
 
+    @Test
+    public void deckPickerNotCrashOnNoPermissionTest() {
+        Application application = ApplicationProvider.getApplicationContext();
+        ShadowApplication app = shadowOf(application);
+        app.denyPermissions(Manifest.permission.READ_EXTERNAL_STORAGE);
+        app.denyPermissions(Manifest.permission.WRITE_EXTERNAL_STORAGE);
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java:269]:
```suggestion
        app.denyPermissions(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE);
```

File: AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java [hunk 2]
```diff
@@ -246,6 +261,25 @@ public void databaseLockedNoPermissionIntegrationTest() {
         }
     }
 
+    @Test
+    public void deckPickerNotCrashOnNoPermissionTest() {
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java:265]: Did this test fail beforehand?

File: AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java [hunk 3]
```diff
@@ -246,6 +261,25 @@ public void databaseLockedNoPermissionIntegrationTest() {
...
+        ShadowApplication app = shadowOf(application);
+        app.denyPermissions(Manifest.permission.READ_EXTERNAL_STORAGE);
+        app.denyPermissions(Manifest.permission.WRITE_EXTERNAL_STORAGE);
+
+        try (ActivityScenario<DeckPicker> scenario = ActivityScenario.launch(DeckPicker.class)) {
+            scenario.onActivity(deckPicker -> {
+                Toolbar toolbar = deckPicker.findViewById(R.id.toolbar);
+                deckPicker.onCreateOptionsMenu(toolbar.getMenu());
+                assertThat("Deck was added", getCol().getDecks().count(), is(1));
+                assertThat("Not visible",toolbar.getMenu().findItem(R.id.deck_picker_action_filter).isVisible(),is(false));
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java:276]: These assertions don't seem relevant to the test

**Comment by david-allison-1** [comment]
> @david-allison-1
> Could you say what might have gone wrong in the approach as one of the test has failed.

Flaky tests, sorry

File: AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java (hunk 3 above)
Comment (shobhi1310) [reply AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java:276]: Actually deck would crash if permissions are not provided because while setting the search function to `isVisible` it will do a `getCol` and that leads to the crash, so I thought using this would simulate the same scenario

File: AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java (hunk 2 above)
Comment (shobhi1310) [reply AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java:265]: No the test did not fail beforehand

Comment (david-allison-1) [reply AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java:265]: To be clear: we need this. 

If this test doesn't fail beforehand, then it wouldn't have picked up a regression, and won't help protect against future bugs

**Comment by Arthur-Milchior** [comment]
There is a conflict with a recently merged PR. Can you please rebase on top of master please?
